File: genna/smbase/smbaseJava/__smbase/State.java
package __smbase;

import java.util.LinkedList;

public abstract class State {
    
    private State directAncestor;
    private LinkedList<Event> waitedTimeEvents;
    protected StateMachine stateMachine;
    
    public State(StateMachine sm){
	directAncestor = null;
	stateMachine = sm;
	waitedTimeEvents = new LinkedList<Event>();
    }
    
    public void setDirectAncestor(State d){
	directAncestor = d;
    }

    public State getDirectAncestor(){
	return directAncestor;
    }
    
    protected void addWaited(Event e){
	waitedTimeEvents.offer(e);
    }

    protected boolean isWaited(Event e){
	return waitedTimeEvents.contains(e);
    }

    protected void removeWaited(Event e){
	waitedTimeEvents.remove(e);
    }

    protected void clearWaited(){
	waitedTimeEvents.clear();
    }

    public abstract EventHandle handleEvent(Event e);
    public void entry() {};
    public void doActivity() {};
    public void exit() {};
}

File: genna/smbase/smbaseCpp/__smbase/State.cpp
#include <algorithm>
#include "State.h"

namespace __smbase{

  State::State(StateMachine *sm){
    directAncestor = 0;
    stateMachine = sm;
  }

  State::~State(){}

  void State::entry(){};

  void State::doActivity(){};

  void State::exit(){};

  void State::setDirectAncestor(State *d){
    directAncestor = d;
  }

  State *State::getDirectAncestor(){
    return directAncestor;
  }

  void State::addWaited(Event *e){
    waitedTimeEvents.push_back(e);
  }

  bool State::isWaited(Event *e){
    std::vector<Event *>::iterator i;

    i = std::find(waitedTimeEvents.begin(), waitedTimeEvents.end(), e);
    
    return i != waitedTimeEvents.end();
  }

  void State::removeWaited(Event *e){
    std::vector<Event *>::iterator i;

    i = std::find(waitedTimeEvents.begin(), waitedTimeEvents.end(), e);

    if(i != waitedTimeEvents.end())
      waitedTimeEvents.erase(i);
  }

  void State::clearWaited(){
    waitedTimeEvents.clear();
  }
}
